### Reading Directories

In [1]:
import os
os.listdir(".\\SpamAssassinMessages")

['easy_ham', 'easy_ham_2', 'hard_ham', 'spam', 'spam_2']

In [2]:
x = os.listdir(".\\SpamAssassinMessages\\easy_ham")


In [3]:
x[0]

'00001.7c53336b37003a9286aba55d2945844c'

### How to open a file the python way

In [4]:
###

with open(os.path.join(".\\SpamAssassinMessages\\easy_ham",x[0]), "r") as file_handler:
    msg = file_handler.read()

### Count the number of files AND can we read them all

In [5]:
count = 0
for msg in x:
    path_ = os.path.join(".\\SpamAssassinMessages\\easy_ham",msg)
    with open(path_,"r") as file_handler:
        msg = file_handler.read()
        if count % 20 == 0:
            print(count)
        count+=1
        

0
20
40
60
80
100
120
140
160
180
200
220
240
260
280
300
320
340
360
380
400
420
440
460
480
500
520
540
560
580
600
620
640
660
680
700
720
740
760
780
800
820
840
860
880
900
920
940
960
980
1000
1020
1040
1060
1080
1100
1120
1140
1160
1180
1200
1220
1240
1260
1280
1300
1320
1340
1360
1380
1400
1420
1440
1460
1480
1500
1520
1540
1560
1580
1600
1620
1640
1660
1680
1700
1720
1740
1760
1780
1800
1820
1840
1860
1880
1900
1920
1940
1960
1980
2000
2020
2040
2060
2080
2100
2120
2140
2160
2180
2200
2220
2240
2260
2280
2300
2320
2340
2360
2380
2400
2420
2440
2460
2480
2500
2520
2540
2560
2580
2600
2620
2640
2660
2680
2700
2720
2740
2760
2780
2800
2820
2840
2860
2880
2900
2920
2940
2960
2980
3000
3020
3040
3060
3080
3100
3120
3140
3160
3180
3200
3220
3240
3260
3280
3300
3320
3340
3360
3380
3400
3420
3440
3460
3480
3500
3520
3540
3560
3580
3600
3620
3640
3660
3680
3700
3720
3740
3760
3780
3800
3820
3840
3860
3880
3900
3920
3940
3960
3980
4000
4020
4040
4060
4080
4100
4120
4140
4160
4180
4200
4

### Get the labels for all our emails

In [6]:
file_name = []
label = []
for root,dirs,files in os.walk(".\\SpamAssassinMessages\\"):
    for f in files:
        if "spam" in root:
            label.append(1)
        else:
            label.append(0)
        file_name.append(os.path.join(root,f))
        

### Create a data frame

In [7]:
import pandas as pd
data = pd.DataFrame({"Message":file_name,"Target":label})

In [8]:
data

Message  Target
0     .\SpamAssassinMessages\easy_ham\00001.7c53336b...       0
1     .\SpamAssassinMessages\easy_ham\00002.9c4069e2...       0
2     .\SpamAssassinMessages\easy_ham\00003.860e3c3c...       0
3     .\SpamAssassinMessages\easy_ham\00004.864220c5...       0
4     .\SpamAssassinMessages\easy_ham\00005.bf27cdea...       0
...                                                 ...     ...
9348  .\SpamAssassinMessages\spam_2\01397.f75f0dd0dd...       1
9349  .\SpamAssassinMessages\spam_2\01398.8ca7045aae...       1
9350  .\SpamAssassinMessages\spam_2\01399.2319643317...       1
9351  .\SpamAssassinMessages\spam_2\01400.b444b69845...       1
9352                 .\SpamAssassinMessages\spam_2\cmds       1

[9353 rows x 2 columns]

### Read in text messages

In [9]:
import email


In [10]:
"""
Lets count the types of messages we have first
"""
from collections import Counter
types = Counter()
msgs = []
trigger = True
for root,dirs,files in os.walk(".\\SpamAssassinMessages\\"):
    for f in files:
        with open(os.path.join(root,f),'r',encoding='latin-1') as file_point:
            msg = email.message_from_file(file_point, )
            type_ = msg.get_content_type()
            types[type_]+=1
            if type_ == 'multipart/mixed' and trigger:
                print(root,f)
                print("______________________")
                trigger = False
                SAMPLE = msg.get_payload()

print(types) 
print("********************************")
print("WARNING--Remember all the multipart (and html!!) messages!!")
print("You need address that for Case Study 3")     

.\SpamAssassinMessages\easy_ham 00067.23813c5ac6ce66fd892ee5501fd5dbd2
______________________
Counter({'text/plain': 7413, 'text/html': 1193, 'multipart/alternative': 326, 'multipart/signed': 180, 'multipart/mixed': 179, 'multipart/related': 56, 'multipart/report': 5, 'text/plain charset=us-ascii': 1})
********************************
WARNING--Remember all the multipart (and html!!) messages!!
You need address that for Case Study 3


In [11]:
"""
Read all the messages in
"""
msgs = []
for root,dirs,files in os.walk("../datasets/SpamAssassinMessages"):
    for f in files:
        with open(os.path.join(root,f),'r',encoding='latin-1') as file_point:
            msg = email.message_from_file(file_point)
            body = msg.get_payload()
            msgs.append(body)

 
print("WARNING--Remember all the multipart messages!!")
print("You need address that for Case Study 3")       

WARNING--Remember all the multipart messages!!
You need address that for Case Study 3


#### Make sure our list of message text is the same as the number of emails

In [12]:
len(msgs)

9353

In [21]:
data['messages'] = msgs


In [22]:
data

Message  Target  \
0     .\SpamAssassinMessages\easy_ham\00001.7c53336b...       0   
1     .\SpamAssassinMessages\easy_ham\00002.9c4069e2...       0   
2     .\SpamAssassinMessages\easy_ham\00003.860e3c3c...       0   
3     .\SpamAssassinMessages\easy_ham\00004.864220c5...       0   
4     .\SpamAssassinMessages\easy_ham\00005.bf27cdea...       0   
...                                                 ...     ...   
9348  .\SpamAssassinMessages\spam_2\01397.f75f0dd0dd...       1   
9349  .\SpamAssassinMessages\spam_2\01398.8ca7045aae...       1   
9350  .\SpamAssassinMessages\spam_2\01399.2319643317...       1   
9351  .\SpamAssassinMessages\spam_2\01400.b444b69845...       1   
9352                 .\SpamAssassinMessages\spam_2\cmds       1   

                                               messages  
0         Date:        Wed, 21 Aug 2002 10:54:46 -05...  
1     Martin A posted:\nTassos Papadopoulos, the Gre...  
2     Man Threatens Explosion In Moscow \n\nThursday...  
3     Klez: The Virus That Won't Die\n \nAlready the...  
4     >  in adding cream to spaghetti carbonara, whi...  
...                                                 ...  
9348  [[Content-Type, Content-Transfer-Encoding], [C...  
9349  Dear Subscriber,\n\nIf I could show you a way ...  
9350  ****Mid-Summer Customer Appreciation SALE!****...  
9351  ATTN:SIR/MADAN      \n\n                      ...  
9352  mv 00001.317e78fa8ee2f54cd4890fdc09ba8176 0000...  

[9353 rows x 3 columns]

In [15]:
### Example of cleaning a message

In [16]:
msgs[1]

"Martin A posted:\nTassos Papadopoulos, the Greek sculptor behind the plan, judged that the\n limestone of Mount Kerdylio, 70 miles east of Salonika and not far from the\n Mount Athos monastic community, was ideal for the patriotic sculpture. \n \n As well as Alexander's granite features, 240 ft high and 170 ft wide, a\n museum, a restored amphitheatre and car park for admiring crowds are\nplanned\n---------------------\nSo is this mountain limestone or granite?\nIf it's limestone, it'll weather pretty fast.\n\n------------------------ Yahoo! Groups Sponsor ---------------------~-->\n4 DVDs Free +s&p Join Now\nhttp://us.click.yahoo.com/pt6YBB/NXiEAA/mG3HAA/7gSolB/TM\n---------------------------------------------------------------------~->\n\nTo unsubscribe from this group, send an email to:\nforteana-unsubscribe@egroups.com\n\n \n\nYour use of Yahoo! Groups is subject to http://docs.yahoo.com/info/terms/ \n\n\n\n"

In [17]:
print(msgs[1].replace("\n"," "))

Martin A posted: Tassos Papadopoulos, the Greek sculptor behind the plan, judged that the  limestone of Mount Kerdylio, 70 miles east of Salonika and not far from the  Mount Athos monastic community, was ideal for the patriotic sculpture.     As well as Alexander's granite features, 240 ft high and 170 ft wide, a  museum, a restored amphitheatre and car park for admiring crowds are planned --------------------- So is this mountain limestone or granite? If it's limestone, it'll weather pretty fast.  ------------------------ Yahoo! Groups Sponsor ---------------------~--> 4 DVDs Free +s&p Join Now http://us.click.yahoo.com/pt6YBB/NXiEAA/mG3HAA/7gSolB/TM ---------------------------------------------------------------------~->  To unsubscribe from this group, send an email to: forteana-unsubscribe@egroups.com     Your use of Yahoo! Groups is subject to http://docs.yahoo.com/info/terms/     


### How a message will be split by CountVectorizer or TfIdf (done for you!)

In [42]:
msgs[0]

'    Date:        Wed, 21 Aug 2002 10:54:46 -0500\n    From:        Chris Garrigues <cwg-dated-1030377287.06fa6d@DeepEddy.Com>\n    Message-ID:  <1029945287.4797.TMDA@deepeddy.vircio.com>\n\n\n  | I can\'t reproduce this error.\n\nFor me it is very repeatable... (like every time, without fail).\n\nThis is the debug log of the pick happening ...\n\n18:19:03 Pick_It {exec pick +inbox -list -lbrace -lbrace -subject ftp -rbrace -rbrace} {4852-4852 -sequence mercury}\n18:19:03 exec pick +inbox -list -lbrace -lbrace -subject ftp -rbrace -rbrace 4852-4852 -sequence mercury\n18:19:04 Ftoc_PickMsgs {{1 hit}}\n18:19:04 Marking 1 hits\n18:19:04 tkerror: syntax error in expression "int ...\n\nNote, if I run the pick command by hand ...\n\ndelta$ pick +inbox -list -lbrace -lbrace -subject ftp -rbrace -rbrace  4852-4852 -sequence mercury\n1 hit\n\nThat\'s where the "1 hit" comes from (obviously).  The version of nmh I\'m\nusing is ...\n\ndelta$ pick -version\npick -- nmh-1.0.4 [compiled on fuchsia.c

### Special tokens often encountered in nlp
These are not used in our CS, FYI only

In [19]:
"<pad>"  # a nothing word so all sentences are same length
"<unk>"  # a rare word or word not in the vocabulart
"<start>" #first word in a sequence

'<start>'

### CountVectorizer Usage

In [20]:
from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer
vectorizer = TfidfVectorizer()
out = vectorizer.fit_transform(data['messages'].astype('str'))

### Sparse Matrix
Out is a sparse matrix (mostly zeroes)
To view it, convert to an array

In [47]:
out

<9353x91258 sparse matrix of type '<class 'numpy.int64'>'
	with 1459773 stored elements in Compressed Sparse Row format>

In [58]:
out[0].toarray().shape

(1, 91258)

### Vocab
There is a dictionary of word to column
Here let's find where the column 'Martin' is counted
Since the default in CountVectorizer is to lower case everything
Martin, martin, MaRtiN are all counted as 'martin'

In [59]:
vectorizer.vocabulary_['martin']

59070

#### How to reverse a dictionary
and translate back and forth between vector and text

In [60]:
reverse = {value:key for key,value in vectorizer.vocabulary_.items()}

In [61]:
out[1].toarray()[:,59070]

array([0.09372373])

Compare these two, Notice what CountVectorizer did and did not include

In [62]:
for i in range(91258):
    if out[0].toarray()[:,i]!=0:
        print(reverse[i],out[0].toarray()[:,i],i)
        
print(msgs[1])

03 [0.06302065] 3416
04 [0.0962336] 3499
0500 [0.04141477] 3599
06fa6d [0.06438832] 3729
10 [0.018492] 5837
1029945287 [0.06438832] 5976
1030377287 [0.06438832] 5989
14 [0.02690246] 6749
17 [0.02858535] 7178
18 [0.12866134] 7310
19 [0.13559448] 7433
2002 [0.0279438] 8034
21 [0.02833562] 8383
46 [0.03372422] 13258
4797 [0.05929191] 13389
4852 [0.37648587] 13442
54 [0.03447056] 14353
55 [0.03358363] 14433
56 [0.03472632] 14523
_______________________________________________ [0.01918134] 20418
able [0.02836295] 21543
actually [0.02627862] 22015
ago [0.03040097] 22718
and [0.01954647] 23600
at [0.01248048] 24844
au [0.03982273] 25022
aug [0.02844557] 25070
been [0.01907218] 26908
both [0.02601528] 28486
by [0.01418776] 29505
can [0.01435291] 30334
chris [0.0352662] 31857
code [0.02630963] 32646
com [0.04235229] 32885
comes [0.0640202] 32922
command [0.10944527] 32944
compiled [0.04002057] 33124
created [0.03462936] 34431
cs [0.04460079] 34687
cvs [0.03607988] 34977
cwg [0.0490991] 35005
da

In [63]:
%%time
from sklearn.naive_bayes import GaussianNB
ng = GaussianNB()
ng.fit(out.toarray(),data['Target'])

CPU times: total: 15.7 s
Wall time: 15.7 s


GaussianNB()

In [64]:
%%time
from sklearn.model_selection import cross_val_score
cross_val_score(ng,out.toarray(),data['Target'],cv=5,n_jobs=1, scoring='accuracy')

CPU times: total: 1min 31s
Wall time: 1min 32s


array([0.9556387 , 0.97755211, 0.87814003, 0.9513369 , 0.81283422])

In [28]:
### Deeper dive into email object

In [33]:
with open(os.path.join(root,li[0]),'r',encoding='latin-1') as file_point:
    msg = email.message_from_file(file_point)  # the message itself
    body = msg.get_payload() # the payload or contents of the message
    """
    The problem we encounter is the 'payload' is not always text
    sometimes the 'payload' is more messages!
    """

In [34]:
msg

In [35]:
msg.get_content_type()

'text/plain'

In [36]:
msg.get_payload()

'    Date:        Wed, 21 Aug 2002 10:54:46 -0500\n    From:        Chris Garrigues <cwg-dated-1030377287.06fa6d@DeepEddy.Com>\n    Message-ID:  <1029945287.4797.TMDA@deepeddy.vircio.com>\n\n\n  | I can\'t reproduce this error.\n\nFor me it is very repeatable... (like every time, without fail).\n\nThis is the debug log of the pick happening ...\n\n18:19:03 Pick_It {exec pick +inbox -list -lbrace -lbrace -subject ftp -rbrace -rbrace} {4852-4852 -sequence mercury}\n18:19:03 exec pick +inbox -list -lbrace -lbrace -subject ftp -rbrace -rbrace 4852-4852 -sequence mercury\n18:19:04 Ftoc_PickMsgs {{1 hit}}\n18:19:04 Marking 1 hits\n18:19:04 tkerror: syntax error in expression "int ...\n\nNote, if I run the pick command by hand ...\n\ndelta$ pick +inbox -list -lbrace -lbrace -subject ftp -rbrace -rbrace  4852-4852 -sequence mercury\n1 hit\n\nThat\'s where the "1 hit" comes from (obviously).  The version of nmh I\'m\nusing is ...\n\ndelta$ pick -version\npick -- nmh-1.0.4 [compiled on fuchsia.c

In [ ]:
### How to explore different content types

In [59]:
from collections import Counter
types = Counter()
msgs = []
for root,dirs,files in os.walk("../datasets/SpamAssassinMessages"):
    for f in files:
        with open(os.path.join(root,f),'r',encoding='latin-1') as file_point:
            msg = email.message_from_file(file_point)
            content = msg.get_content_type()
            payload = msg.get_payload()
            if 'html' in content: # can change
                print(f,root)
                """
                ### Snipper of text to try multiple content emails
                for sub_m in payload:
                    sub_content = sub_m.get_content_type()
                    if sub_content == 'text/plain':
                        print(sub_m.get_payload())
                """
 


00947.5983cb8fa9a4c2a27cb6eca03e0be56d ../datasets/SpamAssassinMessages\easy_ham_2
01318.193fb7308fee59bb4aa70cc72191b0b1 ../datasets/SpamAssassinMessages\easy_ham_2
00007.d24e99a602ee7fb442714c0d448cd08e ../datasets/SpamAssassinMessages\hard_ham
0001.f0cf04027e74802f09f723cb8916b48e ../datasets/SpamAssassinMessages\hard_ham
00011.acdfa5be40e7b6c3ad3df28c63670c7c ../datasets/SpamAssassinMessages\hard_ham
00012.58a866f18474d94989984958e1789df4 ../datasets/SpamAssassinMessages\hard_ham
00015.ada83ed8f5e09b7dd5b268dafb0d7e8d ../datasets/SpamAssassinMessages\hard_ham
00016.47e87c7e7f6c78738ad4fb654dbdaaac ../datasets/SpamAssassinMessages\hard_ham
00018.75bf8472753f24aa22df72c7301e07ec ../datasets/SpamAssassinMessages\hard_ham
00019.e35a7a6a1a6bdd0d2e164db2f6a0e4ef ../datasets/SpamAssassinMessages\hard_ham
0002.2fe846db6e3249836abdbfcae459bf2a ../datasets/SpamAssassinMessages\hard_ham
00020.eca3ca2c144cd6ceaf677f3d8b7b0eed ../datasets/SpamAssassinMessages\hard_ham
00022.66e4bce429ab25c5d2c7

00212.1459a647a827cc74ab6821fb2e9c8c93 ../datasets/SpamAssassinMessages\hard_ham
00214.94d0d7dd09e7b4926f56071a11a2f984 ../datasets/SpamAssassinMessages\hard_ham
00224.0b10471bd4bcd9b5aad01f0431f72d9f ../datasets/SpamAssassinMessages\hard_ham
0023.cb5d5eef370b20c455e62dd60f071fd9 ../datasets/SpamAssassinMessages\hard_ham
0024.aed7657369f5d6cffe71818d06a8de42 ../datasets/SpamAssassinMessages\hard_ham
0025.15d4566a6f33d21e7f2fe5c452e1a9fe ../datasets/SpamAssassinMessages\hard_ham
0027.67c81ed875f5095080aec2de21d50d70 ../datasets/SpamAssassinMessages\hard_ham
0028.fd45453350359f2fca8fbf13ee44bf2d ../datasets/SpamAssassinMessages\hard_ham
0029.cebbcfd6de93eb527880fdc53e71c9f1 ../datasets/SpamAssassinMessages\hard_ham
0030.76c4bd87d3ccc6da6916ec5a5e08800d ../datasets/SpamAssassinMessages\hard_ham
0031.6d5310f0ef66bc099725be92aab80398 ../datasets/SpamAssassinMessages\hard_ham
0032.72c3e5f2fa8153c666ac73ce1fe69301 ../datasets/SpamAssassinMessages\hard_ham
0033.6a01f21844e78eba1202be887c1d9473

00196.dd21040c7757d477c967ae71b537810e ../datasets/SpamAssassinMessages\spam
00199.9be6cec49c53210152780926cdeb59ff ../datasets/SpamAssassinMessages\spam
00200.bacd4b2168049778b480367ca670254f ../datasets/SpamAssassinMessages\spam
00201.00020fc9911604f6cae7ae0f598ad29d ../datasets/SpamAssassinMessages\spam
00202.d5b52386f66bd36cd1508319c82cf671 ../datasets/SpamAssassinMessages\spam
00204.a008813ddeb2d5febd1fc676c07e9760 ../datasets/SpamAssassinMessages\spam
00205.312e72065386636132fa6c4a1fde871e ../datasets/SpamAssassinMessages\spam
00211.d976c6049e8448e7c407c124b580e0ba ../datasets/SpamAssassinMessages\spam
00212.a9947ad74a529a35d11538e1df60cd73 ../datasets/SpamAssassinMessages\spam
00213.8c42a1c257aa30ff3b3ba668cca59408 ../datasets/SpamAssassinMessages\spam
00215.f571ecd203e8d39296419bffc47e4a6a ../datasets/SpamAssassinMessages\spam
00217.43b4ef3d9c56cf42be9c37b546a19e78 ../datasets/SpamAssassinMessages\spam
00221.c4dfeecf0cacc9469540337f5baf69db ../datasets/SpamAssassinMessages\spam

0306.521d917ac6509c499c406647fd0d336b ../datasets/SpamAssassinMessages\spam
0307.2e4dc0cdb1e3b49f0986c19c1f324224 ../datasets/SpamAssassinMessages\spam
0310.23036f6ae05720b052b73117b6ecb957 ../datasets/SpamAssassinMessages\spam
0312.a0e7f2633bd0ceaddf16fba58be54778 ../datasets/SpamAssassinMessages\spam
0314.5b03e0718373f3319eadaec592308aba ../datasets/SpamAssassinMessages\spam
0316.018282fecc304f005e952d7c5c181dad ../datasets/SpamAssassinMessages\spam
0320.e34c9c6f982b8ce353c10aa362d6da17 ../datasets/SpamAssassinMessages\spam
0323.badf0273f656afd0dfebaa63af1c81f6 ../datasets/SpamAssassinMessages\spam
0324.269d2c1815931004c7be2be4d26e8b66 ../datasets/SpamAssassinMessages\spam
0331.1de50a02d91a4e0c6daf5e2cc28a60c6 ../datasets/SpamAssassinMessages\spam
0332.b82bddb316d2e12418d6ea8791ce5896 ../datasets/SpamAssassinMessages\spam
0333.492c1809f9755e305774bed2aaa3aa9b ../datasets/SpamAssassinMessages\spam
0340.8e191c37e2d30a639013203aacf60086 ../datasets/SpamAssassinMessages\spam
0342.babb504

00468.fb2222cc49228bd3dac5481e670f208f ../datasets/SpamAssassinMessages\spam_2
00469.d2c0c55b686454f38eee312c5e190816 ../datasets/SpamAssassinMessages\spam_2
00471.df77fa930951f79466c195052ff56816 ../datasets/SpamAssassinMessages\spam_2
00473.594d47d74b993e949b2b472af3430aed ../datasets/SpamAssassinMessages\spam_2
00475.3d497c7d96c51986316db566756ff35a ../datasets/SpamAssassinMessages\spam_2
00476.38e32c20c334317781e90c3e6754fbe9 ../datasets/SpamAssassinMessages\spam_2
00477.3691f298683e5b8687bc05a891512864 ../datasets/SpamAssassinMessages\spam_2
00478.a4d1f3bcbb571f1c3809bf47cb5ee57f ../datasets/SpamAssassinMessages\spam_2
00479.84c58aa8adff520d7f58dfde02c1cf10 ../datasets/SpamAssassinMessages\spam_2
00480.682829ef5c1f1639f0a060190c5d4b33 ../datasets/SpamAssassinMessages\spam_2
00483.e583ffb0efd3958cdef2e9f3b043ef0d ../datasets/SpamAssassinMessages\spam_2
00485.94b2cb3aa454e6f6701c42cb1fd35ffe ../datasets/SpamAssassinMessages\spam_2
00486.7f5cde6ad9f34dcbe56a1fd73138b351 ../datasets/S

00745.a0f2c78f1a75fe880532f0c432aa12d2 ../datasets/SpamAssassinMessages\spam_2
00749.9887b1d7cb21083c777b0623cfdb02af ../datasets/SpamAssassinMessages\spam_2
00751.3158a29a29997cc16a69497399d90ca2 ../datasets/SpamAssassinMessages\spam_2
00758.13f37fcfbc515a7f7de269852fb7b842 ../datasets/SpamAssassinMessages\spam_2
00761.00d729b279723c9ae9d8f09e171db301 ../datasets/SpamAssassinMessages\spam_2
00762.e31568dff471c947d42869ae2f8f0779 ../datasets/SpamAssassinMessages\spam_2
00763.868a503063713b62fd5325513ba29761 ../datasets/SpamAssassinMessages\spam_2
00764.d81e084a6940b58fa3deabed038a7b9e ../datasets/SpamAssassinMessages\spam_2
00765.cfd85d27a812054ddd5ee1fc7d881557 ../datasets/SpamAssassinMessages\spam_2
00766.ff1f266127aebe1fd285b9a211f34723 ../datasets/SpamAssassinMessages\spam_2
00767.6dc9c38495942ab7a08e9317d14eb77e ../datasets/SpamAssassinMessages\spam_2
00768.9900419b1120aac5256e7b1ba0de2b1f ../datasets/SpamAssassinMessages\spam_2
00769.b4477686a6ab2b52370e3f671ce9a016 ../datasets/S

01033.2af876341f85de2a7f501b8dd0248cd6 ../datasets/SpamAssassinMessages\spam_2
01034.925db77dafa8bdaae1cb31e2ed3b05a4 ../datasets/SpamAssassinMessages\spam_2
01037.3512597a838a3eaf3149924386ef9fd3 ../datasets/SpamAssassinMessages\spam_2
01038.95ce9be665025f2ab826870d509337c6 ../datasets/SpamAssassinMessages\spam_2
01039.40b21f41dcf48f380729c22cd2a62122 ../datasets/SpamAssassinMessages\spam_2
01040.24856bbcaedd4d7b28eae47d8f89a62f ../datasets/SpamAssassinMessages\spam_2
01042.7b53680639a0b4ec9e333ce3046b6af4 ../datasets/SpamAssassinMessages\spam_2
01048.a8581a98e46532472fe74772b6e3477a ../datasets/SpamAssassinMessages\spam_2
01049.621d66148b023203d9010ee5df12ddd1 ../datasets/SpamAssassinMessages\spam_2
01050.f18a04fd3f7cf3e60483c3420bff5417 ../datasets/SpamAssassinMessages\spam_2
01052.223738f00502f5dd7f86dd559af8f795 ../datasets/SpamAssassinMessages\spam_2
01053.2ec9e69df620c85e3c7c6f61ce199a3c ../datasets/SpamAssassinMessages\spam_2
01054.3b393992b2d9b3c2209719e6ac83da11 ../datasets/S

01334.24b7f4702e0da7e9d7a5f4d284adfc96 ../datasets/SpamAssassinMessages\spam_2
01335.ca70ab7f2d1accef8b3cf646e3004ef8 ../datasets/SpamAssassinMessages\spam_2
01336.cce700d418ddce6f5081cdcda625074f ../datasets/SpamAssassinMessages\spam_2
01338.a67c3827402b4610bd2a6814cd8cd907 ../datasets/SpamAssassinMessages\spam_2
01339.d94a16532a6f42497266dcc327a40163 ../datasets/SpamAssassinMessages\spam_2
01341.62703c6047c508d1d84bde0b7653b556 ../datasets/SpamAssassinMessages\spam_2
01343.97e31a95126f0c6dc249a8e51489af10 ../datasets/SpamAssassinMessages\spam_2
01346.fb942e99ad6211fe374675bc9ac639d5 ../datasets/SpamAssassinMessages\spam_2
01347.e2cd456cd2d58601fec5a5b6323463e1 ../datasets/SpamAssassinMessages\spam_2
01348.0ed90bb4a1ba1ea2309ffdbbce093753 ../datasets/SpamAssassinMessages\spam_2
01350.31d85a79d8c3b80069316daf56176820 ../datasets/SpamAssassinMessages\spam_2
01353.369f79f8f31f3b18bdb5d1006207b52e ../datasets/SpamAssassinMessages\spam_2
01354.942feb599d3e244a238c28c9028d97fa ../datasets/S

In [64]:
with open(".\\SpamAssassinMessages\\easy_ham_2\\01318.193fb7308fee59bb4aa70cc72191b0b1",'r',encoding='latin-1') as file_point:
    msg = email.message_from_file(file_point)
    body = msg.get_payload()
    #msgs.append(body)

In [65]:
msg

In [66]:
body

'<html>\n\t<body>\n\t\t<table border="0">\n\t\t\t<tr>\n\t\t\t\t<td width="25">&nbsp;</td>\n\t\t\t\t<td width="450">\n\t\t\t\t\t<table border="0" width="100%">\n\t\t\t\t\t\t<tr>\n\t\t\t\t\t\t\t<td><img src="http://static.userland.com/misc/snImages/scriptingNewsBulletinLogo.gif" width="244" height="36"></td>\n\t\t\t\t\t\t\t<td align="right"><form name="unsubscribeForm" action="http://www.userland.com/dailyUpdatesUnsubscribe" method="POST"><input type="hidden" name="recipient" value="legit-list-scriptingnews@jmason.org"><input type="hidden" name="newsletter" value="scriptingNews"><input type="submit" value="Unsubscribe" name="unsubscribeButton"></form></td>\n\t\t\t\t\t\t\t</tr>\n\t\t\t\t\t\t</table><p>\n\t\t\t\t\t<a name="When:4:29:48PM"><a href="http://radio.weblogs.com/0001015/images/2002/07/30/metsManagementLosers.gif">Apparently</a> the NY Mets couldn\'t find a way to compromise with some of the team\'s most dedicated fans. A sad day for New York baseball. What would Mookie say??&nbsp

In [54]:
for i in body:
    print(i.get_content_type())

text/plain
application/pgp-signature


### How to add a column to your CountVectorize/TfIDF

In [43]:
f=out.toarray()

In [44]:
f.shape

(9353, 91258)

In [45]:
import numpy as np
# arrays aren't smart, so you have to add a column and now a row
# reshape(-1,1) means "put all the numbers in a matrix with 1 column, and as many rows as needed"
# "-1" means "as many rows as needed"
# reshape(1,-1) means "1 row, as many columns as needed"
f = np.append(f,np.zeros(9353).reshape(-1,1),axis=1)

In [46]:
f.shape

(9353, 91259)

### What happens 
when you don't read the email correctly (aka it wasn't 'text/plain')
Look entry index 1040

In [114]:
data[data["Message"].str.contains('00947')]

Message  Target  \
1040  .\SpamAssassinMessages\easy_ham\00947.488310e3...       0   
5998  .\SpamAssassinMessages\easy_ham_2\00947.5983cb...       0   
8898  .\SpamAssassinMessages\spam_2\00947.e4d80402e9...       1   

                                               messages  
1040                   [[Content-Type], [Content-Type]]  
5998  \nIt took me a week to get down to this;\nThe ...  
8898  <HR>\n<html>\n<div bgcolor=3D"#FFFFCC">\n\n  <...

### Text Cleaning Help
You can also use regex, Beuatiful Soup, whatever tools you need.

In [47]:
"this"
"this;"
"this,".replace(",","")

'this'

In [48]:
x="This is my big, long, dumb, sentance!"
words = x.split(" ")

In [49]:
words

['This', 'is', 'my', 'big,', 'long,', 'dumb,', 'sentance!']

In [50]:
my_list = [",","!"]
for item in my_list:
    words = [word.replace(item,"") for word in words]

In [51]:
words

['This', 'is', 'my', 'big', 'long', 'dumb', 'sentance']

In [25]:
with open("./SpamAssassinMessages/easy_ham/00067.23813c5ac6ce66fd892ee5501fd5dbd2") as f:
    msg = email.message_from_file(f)

In [26]:
msg.get_payload()

In [38]:
def return_txt(msg):
    if msg.get_content_type()=='text/plain':
        return msg.get_payload()
    if msg.get_content_type()=='text/html':
        return "I BETTER ADD SOME CODE HERE"
    else:
        for item in msg.get_payload():
            return return_txt(item)
            break



In [39]:
x=return_txt(msg)

In [41]:
data['message']

KeyError: 'message'

In [45]:
text = []
for root,dirs,files in os.walk("../datasets/SpamAssassinMessages"):
    for f in files:
        with open(os.path.join(root,f),'r',encoding='latin-1') as file_point:
            msg = email.message_from_file(file_point)
            try:
                text.append(return_txt(msg))
            except:
                print(f"{root,f} failed to parse")
            

('../datasets/SpamAssassinMessages\\spam', '00467.5b733c506b7165424a0d4a298e67970f') failed to parse
('../datasets/SpamAssassinMessages\\spam', '0478.16f274ffd3e884a574c9461012cadea6') failed to parse
('../datasets/SpamAssassinMessages\\spam_2', '00204.4cf15f97b8ea08bfafab7d5091b8fbe7') failed to parse
('../datasets/SpamAssassinMessages\\spam_2', '01214.973b4598b630a989967ff69b19f95d4a') failed to parse
